File: frontend/src/lib/srp.ts
import { fromHex, randomBytes, sha256, toHex, utf8, type Bytes } from './bytes'

const N_HEX =
  'ac6bdb41324a9a9bf166de5e1389582faf72b6651987ee07fc3192943db56050' +
  'a37329cbb4a099ed8193e0757767a13dd52312ab4b03310dcd7f48a9da04fd50' +
  'e8083969edb767b0cf6095179a163ab3661a05fbd5faaae82918a9962f0b93b8' +
  '55f97993ec975eeaa80d740adbf4ff747359d041d5c33ea71d281e446b14773b' +
  'ca97b43a23fb801676bd207a436c6481f1d2b9078717461a5b9d32e688f87748' +
  '544523b524b0d57d5ea77a2775d2ecfa032cfbdbf52fb3786160279004e57ae6' +
  'af874e7303ce53299ccc041c7bc308d82a5698f3a8d0c38271ae35f8e9dbfbb6' +
  '94b5c803d89f7ae435de236d525f54759b65e372fcd68ef20fa7111f9e4aff73'

const N = BigInt('0x' + N_HEX)
const G = 2n
const KEY_LENGTH = 256

function intFromHex(hex: string): bigint {
  const value = BigInt('0x' + hex)
  if (value < 0n || value >= N) throw new Error('value out of group range')
  return value
}

function bigToBytes(value: bigint, length: number = KEY_LENGTH): Bytes {
  const out = new Uint8Array(length)
  let current = value
  for (let i = length - 1; i >= 0; i--) {
    out[i] = Number(current & 0xffn)
    current >>= 8n
  }
  return out
}

function bytesToBig(bytes: Bytes): bigint {
  let value = 0n
  for (const byte of bytes) value = (value << 8n) | BigInt(byte)
  return value
}

export function hexFromInt(value: bigint): string {
  return value.toString(16).padStart(KEY_LENGTH * 2, '0')
}

async function pad(value: bigint): Promise<Bytes> {
  return bigToBytes(value)
}

async function multiplierK(): Promise<bigint> {
  return bytesToBig(await sha256(bigToBytes(N), bigToBytes(G)))
}

export async function derivePrivateKeyKey(saltHex: string, username: string, password: string): Promise<bigint> {
  const inner = await sha256(utf8(`${username}:${password}`))
  return bytesToBig(await sha256(fromHex(saltHex), inner))
}

export async function deriveVerifier(saltHex: string, username: string, password: string): Promise<string> {
  const x = await derivePrivateKeyKey(saltHex, username, password)
  return hexFromInt(modPow(G, x, N))
}

export function randomPrivate(): bigint {
  return bytesToBig(randomBytes(64)) % (N - 1n) + 1n
}

export async function clientEphemeral(): Promise<{ private: bigint; publicHex: string }> {
  const a = randomPrivate()
  return { private: a, publicHex: hexFromInt(modPow(G, a, N)) }
}

export interface SrpClientSession {
  keyHex: string
  clientProofHex: string
  serverProofHex: string
}

export async function clientSession(
  clientPrivate: bigint,
  serverPublicHex: string,
  saltHex: string,
  username: string,
  password: string,
): Promise<SrpClientSession> {
  const B = intFromHex(serverPublicHex)
  if (B % N === 0n) throw new Error('invalid server ephemeral')
  const A = modPow(G, clientPrivate, N)
  const u = bytesToBig(await sha256(await pad(A), await pad(B)))
  if (u === 0n) throw new Error('invalid scrambling parameter')
  const x = await derivePrivateKeyKey(saltHex, username, password)
  const k = await multiplierK()
  const base = ((B - k * modPow(G, x, N)) % N + N) % N
  const S = modPow(base, clientPrivate + u * x, N)
  const key = await sha256(await pad(S))
  const clientProof = await sha256(await pad(A), await pad(B), key)
  const serverProof = await sha256(await pad(A), clientProof, key)
  return {
    keyHex: toHex(key),
    clientProofHex: toHex(clientProof),
    serverProofHex: toHex(serverProof),
  }
}

function modPow(base: bigint, exponent: bigint, modulus: bigint): bigint {
  let result = 1n
  let b = ((base % modulus) + modulus) % modulus
  let e = exponent
  while (e > 0n) {
    if (e & 1n) result = (result * b) % modulus
    b = (b * b) % modulus
    e >>= 1n
  }
  return result
}

export { N, G }

File: backend/apps/common/srp.py
import hashlib
import hmac
import secrets

N = int(
    "AC6BDB41324A9A9BF166DE5E1389582FAF72B6651987EE07FC3192943DB56050"
    "A37329CBB4A099ED8193E0757767A13DD52312AB4B03310DCD7F48A9DA04FD5"
    "0E8083969EDB767B0CF6095179A163AB3661A05FBD5FAAAE82918A9962F0B93"
    "B855F97993EC975EEAA80D740ADBF4FF747359D041D5C33EA71D281E446B147"
    "73BCA97B43A23FB801676BD207A436C6481F1D2B9078717461A5B9D32E688F8"
    "7748544523B524B0D57D5EA77A2775D2ECFA032CFBDBF52FB3786160279004E"
    "57AE6AF874E7303CE53299CCC041C7BC308D82A5698F3A8D0C38271AE35F8E9D"
    "BFBB694B5C803D89F7AE435DE236D525F54759B65E372FCD68EF20FA7111F9E4AFF73",
    16,
)
G = 2
KEY_LENGTH = 256


def _hash(*parts):
    hasher = hashlib.sha256()
    for part in parts:
        hasher.update(part)
    return hasher.digest()


def _pad(value):
    return value.to_bytes(KEY_LENGTH, "big")


def int_from_hex(value):
    return int(value, 16)


def hex_from_int(value):
    return format(value, "0{}x".format(KEY_LENGTH * 2))


def multiplier_k():
    return int.from_bytes(_hash(_pad(N), _pad(G)), "big")


def generate_private_ephemeral():
    return secrets.randbelow(N - 1) + 1


def generate_server_ephemeral(verifier):
    private = generate_private_ephemeral()
    public = (multiplier_k() * verifier + pow(G, private, N)) % N
    return private, public


def derive_private_key(salt_hex, username, password):
    inner = hashlib.sha256("{}:{}".format(username, password).encode()).digest()
    return int.from_bytes(_hash(bytes.fromhex(salt_hex), inner), "big")


def derive_verifier(private_key):
    return pow(G, private_key, N)


def _scrambling(public_a, public_b):
    u = int.from_bytes(_hash(_pad(public_a), _pad(public_b)), "big")
    if u == 0:
        raise ValueError("invalid scrambling parameter")
    return u


def server_session(server_private, client_public, verifier):
    if client_public % N == 0:
        raise ValueError("invalid client ephemeral")
    server_public = (multiplier_k() * verifier + pow(G, server_private, N)) % N
    u = _scrambling(client_public, server_public)
    shared = pow(client_public * pow(verifier, u, N), server_private, N)
    key = _hash(_pad(shared))
    client_proof = _hash(_pad(client_public), _pad(server_public), key)
    server_proof = _hash(_pad(client_public), client_proof, key)
    return {
        "key": key.hex(),
        "client_proof": client_proof.hex(),
        "server_proof": server_proof.hex(),
    }


def client_session(client_private, server_public, salt_hex, username, password):
    if server_public % N == 0:
        raise ValueError("invalid server ephemeral")
    client_public = pow(G, client_private, N)
    u = _scrambling(client_public, server_public)
    private_key = derive_private_key(salt_hex, username, password)
    base = (server_public - multiplier_k() * pow(G, private_key, N)) % N
    shared = pow(base, client_private + u * private_key, N)
    key = _hash(_pad(shared))
    client_proof = _hash(_pad(client_public), _pad(server_public), key)
    server_proof = _hash(_pad(client_public), client_proof, key)
    return {
        "key": key.hex(),
        "client_proof": client_proof.hex(),
        "server_proof": server_proof.hex(),
    }


def proofs_match(expected_hex, provided_hex):
    expected = expected_hex.strip().lower().encode()
    provided = provided_hex.strip().lower().encode()
    return hmac.compare_digest(expected, provided)
